src/core/types/ui_settings.ts

```diff
@@ -80,6 +81,13 @@ export interface UiSettingsParams<T = unknown> {
    * Used to validate value on write and read.
    */
   schema: Type<T>;
+  /**
+   * Metric to track once this property changes
+   */
+  metric?: {
+    type: UiStatsMetricType;
+    name: string;
+  };
```

Comment:
What bothers me with this approach is that the responsibilities around this new feature are, imho, wrongly dispatched:

We are defining the metric metadata for a given uiSetting on `UiSettingsParams`, and yet, `core` that is the 'owner' of this type / functionality will not uses at all. Instead, that's the responsibility of the `advancedSettings` plugin to use this new `metric` configuration. If this configuration is registered from core APIs, I would expect the `uiSettingsClient` to have the responsibility to perform the call `usageCollection.reportUiStats` when updating a setting.

However, I know that this is not currently possible, as `core` has no way to access usage collection, which is currently a plugin, so I don't really have a better solution that wouldn't include creating an additional hook in core to allow registering usage collector in it, as we decided to exclude this solution in some previous discussions.

cc @elastic/kibana-telemetry as that reminds me of the issue about moving usage collection to a core service
cc @restrry what's your opinion on this?

Reply:
+1 for moving the `usageCollector` to core. I also think now is a good time to do so.

Reply:
We introduced `CoreUsageDataService` in https://github.com/elastic/kibana/pull/79101 so core can collect usage data now.

Reply:
It can collect data, but I'm not sure we can perform the required calls to `reportUiStats` atm ?

Reply:
Yes, it's only server-side. Though in this case I think doing this server-side is preferable so that we also pick up on `uiSettings.overrides` set in kibana.yml. 

As an aside, we could use the same approach to implement a browser-side `CoreUsageDataService`. I'm neutral on whether telemetry should be inside Core, but I just don't think we have to block any use cases on that decision.

Reply:
I'm a little bit unsure if we want to centralize or distribute this kind of telemetry should core report this, or the plugin that owns the setting? We could report telemetry on all UISettings but that will increase the telemetry payload significantly. The other option is for a plugin to just read a uisetting and do it's own usage collection reporting on that setting. Unless we feel like most plugins will want telemetry on most of their UiSettings maybe it's simpler if the plugin that cares just reports this in it's own usage data?

Reply:
> maybe it's simpler if the plugin that cares just reports this in it's own usage data?

++ here. There are so many ways to change uiSettings: UI, REST API call, kibana.yml. Plugin reading from UiSettings service and reporting set values seems to be the most reliable way to implement this.

Reply:
> maybe it's simpler if the plugin that cares just reports this in it's own usage data?

Which is what this PR does, isn't it? So we're good with the implementation?

Reply:
> Which is what this PR does, isn't it? So we're good with the implementation?

My initial suggestion was similar to https://github.com/elastic/kibana/pull/82860/files#r519857620 
A plugin reads its UiSettings value and reports them. But it seems that in this case, we want to count how many times the value's been toggled.
Read UiSettings and compare it with the last stored value might not be an option since multiple Kibana users would report the "toggle" event.
The current solution works well until we want to track something more complex than `boolean` value. For example, for an array, one might want to track: the number of inserts, number of array items, number of permutations, etc. 

What if we track what a plugin registers every "UiSetting" and inform a plugin-owner about `write` operation to implement a custom logic. As a bonus, we could restrict access to other plugin settings.

Reply:
@rudolf in theory, we could leverage `CoreUsageDataService` to report this `count` stats exposed by UiSettings service, but the core should be able to filter out keys then, which it doesn't know. We can merge the current solution considering it to-be-refactored-soon. WDYT?

Reply:
yeah, as an alternative to https://github.com/elastic/kibana/issues/83084 we could implement UiSettings telemetry inside `CoreUsageDataService`. My concern is that we implement something which only solves one use case but doesn't work for other plugins. But it could very well be that if we just track UiSetting toggle events this is good enough for 90% of our telemetry needs. In such a case we just track a toggle event whether it's a boolean, array, string etc. If we go down this route we should just get some feedback from other plugins to validate that this solves their needs.

Though in general it feels like we might end up with a better design if we leave this outside of core and if it becomes adopted by many solutions and stabilises we pull it into core.

So I would say let's merge this and see refactor it later :+1: